src/text/raw_text_writer.rs

```diff
@@ -81,27 +188,68 @@ impl<W: Write> RawTextWriter<W> {
         self.output.get_mut()
     }
 
-    /// Returns true if the TextWriter is currently positioned within a Struct.
+    /// Returns true if the RawTextWriter is currently positioned within a Struct.
     pub fn is_in_struct(&self) -> bool {
-        if let Some(IonType::Struct) = self.containers.last() {
-            return true;
-        }
-        false
+        self.parent_level().container_type == ContainerType::Struct
+    }
+
+    /// Returns the number of values that have already been written in this container.
+    /// Before the first value in a container is written, this returns `0`.
+    /// For the purposes of this method, the top level is considered a container.
+    fn index_within_parent(&self) -> usize {
+        self.parent_level().child_count
     }
 
-    // Called after each value is written to emit an appropriate delimiter before the next value.
+    /// Returns the `&EncodingLevel` into which the RawTextWriter most recently stepped.
+    fn parent_level(&self) -> &EncodingLevel {
+        // `self.containers` is never empty; it always has at least the top level.
+        self.containers.last().unwrap()
+    }
+
+    /// Increments the value returned by [Self::index_within_parent].
+    fn increment_child_count(&mut self) {
+        let parent_level = self.containers.last_mut().unwrap();
+        parent_level.child_count += 1;
+    }
+
+    /// Called after each value is written to emit an appropriate delimiter before the next value.
     fn write_value_delimiter(&mut self) -> IonResult<()> {
-        use IonType::*;
-        let delimiter = match self.containers.last() {
-            Some(Struct) | Some(List) => ",",
-            Some(SExpression) => " ",
-            Some(scalar) => unreachable!("Inside a non-container type: {:?}", scalar),
-            None => "\n", // Top-level values appear on their own line
+        use ContainerType::*;
+        let delimiter = match self.parent_level().container_type {
+            TopLevel => "",
+            Struct | List => ",",
+            SExpression => "",
```

Comment:
🗺️ Using a `ContainerType` instead of `Option<IonType>` means that I don't have to handle `unreachable!` arms of the `match` any more.